kratos/processes/reorder_and_optimize_modelpart_process.cpp

```diff
@@ -0,0 +1,257 @@
+//                             KRATOS_ERROR << (*it) << "node in condition " << (*it)->GetGeometry()[k].Id() << " is not in the main model part" << std::endl;
+//                         }
+//                     }
+//                 }
+//             }
+            
+            
+            
+
+
+            KRATOS_CATCH("")
+            
+        }
+        
+        
+        void ReorderAndOptimizeModelPartProcess::ActualizeSubModelPart(ModelPart& subpart)
+        {
+            //make a parallel clone of all the nodes
+            #pragma omp parallel for
+            for(int i=0; i<static_cast<int>(subpart.Nodes().size()); ++i)
+            {
+                Node<3>::Pointer& pnode = *(subpart.NodesBegin() + i).base();
+                pnode = mrModelPart.Nodes()(pnode->Id());
+            }
+            subpart.Nodes().Sort();
+
+            #pragma omp parallel for
+            for(int i=0; i<static_cast<int>(subpart.Elements().size()); ++i)
+            {
+                Element::Pointer& pelem = *(subpart.ElementsBegin() + i).base();  
+                pelem = mrModelPart.Elements()(pelem->Id());
+
+            }
+
+            #pragma omp parallel for
+            for(int i=0; i<static_cast<int>(subpart.Conditions().size()); ++i)
+            {
+                Condition::Pointer& pcond = *(subpart.ConditionsBegin() + i).base();            
+                pcond = mrModelPart.Conditions()(pcond->Id());
+            }
+            
+             //actualize pointers within submodelparts
+            for(auto& part : subpart.SubModelParts())
+            {
+                ActualizeSubModelPart(part);
+            }                      
+        }
+        
+        void ReorderAndOptimizeModelPartProcess::OptimizeOrdering() 
+        {
+            const unsigned int n = mrModelPart.Nodes().size();
+            std::vector< std::set< std::size_t > > graph(n);
+            
+            for(int i=0; i<static_cast<int>(mrModelPart.Elements().size()); ++i)
+            {
+                const auto& elem = mrModelPart.ElementsBegin() + i;         
+                const auto& geom = elem->GetGeometry();
+                for(unsigned int k=0; k<geom.size(); ++k)
+                    for(unsigned int l=0; l<geom.size(); ++l)
+                        graph[geom[k].Id()-1].insert(geom[l].Id()-1); //the -1 is because the ids of our nodes start in 1
+            }
+            
+            for(int i=0; i<static_cast<int>(mrModelPart.Conditions().size()); ++i)
+            {
+                const auto& cond = mrModelPart.ConditionsBegin() + i;              
+                const auto& geom = cond->GetGeometry();
+                for(unsigned int k=0; k<geom.size(); ++k)
+                    for(unsigned int l=0; l<geom.size(); ++l)
+                        graph[geom[k].Id()-1].insert(geom[l].Id()-1); //the -1 is because the ids of our nodes start in 1
+            }
+            
+             
+            unsigned int nnz = 0;
```

Comment:
nnz = Number of Non Zeros. I think it is commonly called nnz in this context (at least within ublas)